In [1]:
import pandas as pd # import pandas library for data manipulation and analysis
from pathlib import Path

import matplotlib.pyplot as plt
import seaborn as sns

# Use Path.cwd().parent to go up 1 folder level
BASE_DIR = Path.cwd().parent
DATA_PATH = BASE_DIR / "data" / "WA_Fn-UseC_-Telco-Customer-Churn.csv"

#print(BASE_DIR)
Telco_cust_churn_data = pd.read_csv(DATA_PATH) # read the CSV file into a pandas DataFrame




In [7]:
Telco_cust_churn_data.head() # display the first few rows of the DataFrame

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


In [8]:
Telco_cust_churn_data.info() # display information about the DataFrame, including column names, data types, and non-null counts

<class 'pandas.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   str    
 1   gender            7043 non-null   str    
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   str    
 4   Dependents        7043 non-null   str    
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   str    
 7   MultipleLines     7043 non-null   str    
 8   InternetService   7043 non-null   str    
 9   OnlineSecurity    7043 non-null   str    
 10  OnlineBackup      7043 non-null   str    
 11  DeviceProtection  7043 non-null   str    
 12  TechSupport       7043 non-null   str    
 13  StreamingTV       7043 non-null   str    
 14  StreamingMovies   7043 non-null   str    
 15  Contract          7043 non-null   str    
 16  PaperlessBilling  7043 non-null   str    
 17  Paymen

In [9]:
Telco_cust_churn_data.shape # display the shape of the DataFrame (number of rows and columns)

(7043, 21)

In [10]:
Telco_cust_churn_data.dtypes

customerID              str
gender                  str
SeniorCitizen         int64
Partner                 str
Dependents              str
tenure                int64
PhoneService            str
MultipleLines           str
InternetService         str
OnlineSecurity          str
OnlineBackup            str
DeviceProtection        str
TechSupport             str
StreamingTV             str
StreamingMovies         str
Contract                str
PaperlessBilling        str
PaymentMethod           str
MonthlyCharges      float64
TotalCharges            str
Churn                   str
dtype: object

In [11]:
Telco_cust_churn_data.describe() # display summary statistics for numerical columns in the DataFrame

,SeniorCitizen,tenure,MonthlyCharges
count,7043.000000,7043.000000,7043.000000
mean,0.162147,32.371149,64.761692
std,0.368612,24.559481,30.090047
min,0.000000,0.000000,18.250000
25%,0.000000,9.000000,35.500000
50%,0.000000,29.000000,70.350000
75%,0.000000,55.000000,89.850000
max,1.000000,72.000000,118.750000


# summary statistics for numerical columns for total 7043 observation
## SeniorCitizen
### 1] 16% observation are senior citizen
### 2] as it is binary value only 0/1 is min/max

## tenure
### 1] customer stays witth company is 32.37 ~2.7 year
### 2] min is 0 means some customer are new
### 3] max is 72 , some customer stays for long duration ~ 6 year
### 4] 9 , 29, 55 --> 25 percentile are for < 1 year while 50 percentile stay with < 2 year 5 month even 75% intouch < 4 year 7 month

## MonthlyCharges
### 1] avg unit amount is 64.76
### 2] min / max is 18.25 / 118.750000 unit charges per month
### 3] 25%, 50%, 70% --> 35, 70, 89 pay for month

## as std value
### 1] SeniorCitizen has 0.3686 low spread near 0
### 2] tenure 24.56 month < mean 32.37 , diverse customer lifespans
### 3] MonthlyCharges has 30 big variation in billing amounts as far from mean 64.76

In [12]:
Telco_cust_churn_data.nunique() # display the number of unique values in each column of the DataFrame

customerID          7043
gender                 2
SeniorCitizen          2
Partner                2
Dependents             2
tenure                73
PhoneService           2
MultipleLines          3
InternetService        3
OnlineSecurity         3
OnlineBackup           3
DeviceProtection       3
TechSupport            3
StreamingTV            3
StreamingMovies        3
Contract               3
PaperlessBilling       2
PaymentMethod          4
MonthlyCharges      1585
TotalCharges        6531
Churn                  2
dtype: int64

In [16]:
### Missing Values Check
Telco_cust_churn_data.isnull().sum() # check for missing values in each column of the DataFrame

#as ouutput no null values found in any column of the dataset, indicating that there are no missing values to handle.

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

In [18]:
# Duplicate Check
Telco_cust_churn_data.duplicated().sum() # check for duplicate rows in the DataFrame

# as again no duplicate rows found in the dataset, indicating that there are no duplicate entries to handle.

np.int64(0)

In [19]:
# Distribution of single variables
# sns.histplot()


In [2]:


# Create output directory for plots
output_dir = BASE_DIR / "plots"
output_dir.mkdir(exist_ok=True)

# Find numerical columns
numerical_cols = Telco_cust_churn_data.select_dtypes(include=['int64', 'float64']).columns.tolist()
print(f"Numerical columns: {numerical_cols}")

# Create histplots for each numerical column
for col in numerical_cols:
    plt.figure(figsize=(10, 5))
    sns.histplot(data=Telco_cust_churn_data, x=col, kde=True, bins=30)
    plt.title(f'Distribution of {col}')
    plt.xlabel(col)
    plt.ylabel('Frequency')
    
    # Save the plot
    plot_path = output_dir / f"{col}_histplot.png"
    plt.savefig(plot_path, dpi=300, bbox_inches='tight')
    print(f"Saved: {plot_path}")
    plt.close()

print(f"\nAll histplots saved to: {output_dir}")


Numerical columns: ['SeniorCitizen', 'tenure', 'MonthlyCharges']
Saved: c:\Users\pradi\Pradip Workspace\The IOT acadamy assignment\AI ML Internship Programe\Telco Customer Churn\plots\SeniorCitizen_histplot.png
Saved: c:\Users\pradi\Pradip Workspace\The IOT acadamy assignment\AI ML Internship Programe\Telco Customer Churn\plots\tenure_histplot.png
Saved: c:\Users\pradi\Pradip Workspace\The IOT acadamy assignment\AI ML Internship Programe\Telco Customer Churn\plots\MonthlyCharges_histplot.png

All histplots saved to: c:\Users\pradi\Pradip Workspace\The IOT acadamy assignment\AI ML Internship Programe\Telco Customer Churn\plots


In [4]:
# Find non-numerical (categorical) columns
categorical_cols = Telco_cust_churn_data.select_dtypes(include=['object']).columns.tolist()

# Filter out ID columns and columns with too many unique values
columns_to_plot = [col for col in categorical_cols 
                   if col not in ['customerID'] 
                   and Telco_cust_churn_data[col].nunique() < 20]

print(f"Categorical columns to visualize: {columns_to_plot}\n")

# Create countplots for each categorical column
for col in columns_to_plot:
    num_categories = Telco_cust_churn_data[col].nunique()
    fig_width = max(8, num_categories * 0.6)
    
    plt.figure(figsize=(fig_width, 5))
    sns.countplot(data=Telco_cust_churn_data, x=col, order=Telco_cust_churn_data[col].value_counts().index)
    plt.title(f'Distribution of {col}')
    plt.xlabel(col)
    plt.ylabel('Count')
    plt.xticks(rotation=45, ha='right')
    plt.tight_layout()
    
    # Save the plot
    plot_path = output_dir / f"{col}_countplot.png"
    plt.savefig(plot_path, dpi=100, bbox_inches='tight')
    print(f"Saved: {plot_path}")
    plt.close()

print(f"\nAll categorical countplots saved to: {output_dir}")


C:\Users\pradi\AppData\Local\Temp\ipykernel_20760\10424999.py:2: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  categorical_cols = Telco_cust_churn_data.select_dtypes(include=['object']).columns.tolist()


Categorical columns to visualize: ['gender', 'Partner', 'Dependents', 'PhoneService', 'MultipleLines', 'InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling', 'PaymentMethod', 'Churn']

Saved: c:\Users\pradi\Pradip Workspace\The IOT acadamy assignment\AI ML Internship Programe\Telco Customer Churn\plots\gender_countplot.png
Saved: c:\Users\pradi\Pradip Workspace\The IOT acadamy assignment\AI ML Internship Programe\Telco Customer Churn\plots\Partner_countplot.png
Saved: c:\Users\pradi\Pradip Workspace\The IOT acadamy assignment\AI ML Internship Programe\Telco Customer Churn\plots\Dependents_countplot.png
Saved: c:\Users\pradi\Pradip Workspace\The IOT acadamy assignment\AI ML Internship Programe\Telco Customer Churn\plots\PhoneService_countplot.png
Saved: c:\Users\pradi\Pradip Workspace\The IOT acadamy assignment\AI ML Internship Programe\Telco Customer Churn\plots\MultipleLines_countplot.pn